[Jason Ciu Putra Sung] iP

JasonCP14 opened the pull request:
# Bocil

> “The task of us is never as great as the power behind us.” – Ralph Waldo Emerson ([source](https://statustown.com/quote/5659/#:~:text=The%20task%20ahead%20of%20us,%2D%20Ralph%20Waldo%20Emerson%20quotes))

Bocil frees your mind of having to remember things you need to do, while ensuring your convenience in using it. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. Download it from [here](https://github.com/JasonCP14/ip/releases).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [X] Managing tasks
- [X] Managing deadlines
- [X] Finding spefic task names
- [ ] Reminders (coming soon)

JasonCP14 committed:
Added level 1 functionality

JasonCP14 committed:
Added level 2 functionality

JasonCP14 committed:
Added level 3 functionality

JasonCP14 committed:
Added level 4 functionality

JasonCP14 committed:
Added TaskList Class

JasonCP14 committed:
Added level 5 functionality

JasonCP14 committed:
Added level 6 functionality

JasonCP14 committed:
Removed redundant statements

JasonCP14 committed:
Added text UI testing

JasonCP14 committed:
Add level 7 functionality

JasonCP14 committed:
Add level 8 functionality

JasonCP14 committed:
Add function to parse time

JasonCP14 committed:
Merge branch-level-7 and branch-level-8

JasonCP14 committed:
Separate into new classes and update text UI testing

JasonCP14 committed:
Divide classes into packages

JasonCP14 committed:
Add gradle files

JasonCP14 committed:
Add JUnit tests

JasonCP14 committed:
Remove UiTest

JasonCP14 committed:
Modify gradle dependencies

JasonCP14 committed:
Add JavaDoc comments to the classes and methods

JasonCP14 committed:
Edit the code to follow coding standard

JasonCP14 committed:
Add level 9 functionality

JasonCP14 committed:
Merge branch 'branch-A-CodingStandard'

JasonCP14 committed:
Merge branch-A-JavaDoc, branch-A-CodingStandard and branch-Level-9

JasonCP14 committed:
Add more exception handlings

JasonCP14 committed:
Change class variable access modifier to private

JasonCP14 committed:
Fix coding standard violations

JasonCP14 committed:
Add Executor class

JasonCP14 committed:
Add JavaDocs for DukeException

deeyonn commented on build.gradle:
```diff
@@ -0,0 +1,66 @@
+    archiveClassifier = null
+}
+
+checkstyle {
+    toolVersion = '10.2'
+}
```
This change will stop checkstyle from checking `src/test`
```suggestion
checkstyle {
    toolVersion = '10.2'
    sourceSets  = [project.sourceSets.main]
}
```

JasonCP14 replied:
Thanks for the comments!

deeyonn commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+     * Accepts user input and prints the response.
+     */
+    public void run() {
+        while (true) {
```
I think its better if you can extract a class Command that returns `boolean isExit = True` when the command parsed is an exit command.
Here are some [references](https://nus-cs2103-ay2223s1.github.io/website/schedule/week3/project.html#a-moreoop) under **Stretch Goal**

JasonCP14 committed:
Change class name to Bocil and BocilException

deeyonn commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,127 @@
+public class Parser {
+
+    /** Date patterns accepted by the parser. */
+    private static final String[] DATE_PATTERNS = {
+        "yyyy-M-d HH:mm", "yyyy/M/d HH:mm", "yyyy M d HH:mm",
+        "d-M-yyyy HH:mm", "d/M/yyyy HH:mm", "d M yyyy HH:mm",
+        "yyyy-MM-d HH:mm", "yyyy/MM/d HH:mm", "yyyy MM d HH:mm",
+        "d-MM-yyyy HH:mm", "d/MM/yyyy HH:mm", "d MM yyyy HH:mm",
+        "yyyy-MMM-d HH:mm", "yyyy/MMM/d HH:mm", "yyyy MMM d HH:mm",
+        "d-MMM-yyyy HH:mm", "d/MMM/yyyy HH:mm", "d MMM yyyy HH:mm",
+        "yyyy-MMMM-d HH:mm", "yyyy/MMMM/d HH:mm", "yyyy MMMM d HH:mm",
+        "d-MMMM-yyyy HH:mm", "d/MMMM/yyyy HH:mm", "d MMMM yyyy HH:mm",
+        "yyyy-M-dd HH:mm", "yyyy/M/dd HH:mm", "yyyy M dd HH:mm",
+        "dd-M-yyyy HH:mm", "dd/M/yyyy HH:mm", "dd M yyyy HH:mm",
+        "yyyy-MM-dd HH:mm", "yyyy/MM/dd HH:mm", "yyyy MM dd HH:mm",
+        "dd-MM-yyyy HH:mm", "dd/MM/yyyy HH:mm", "dd MM yyyy HH:mm",
+        "yyyy-MMM-dd HH:mm", "yyyy/MMM/dd HH:mm", "yyyy MMM dd HH:mm",
+        "dd-MMM-yyyy HH:mm", "dd/MMM/yyyy HH:mm", "dd MMM yyyy HH:mm",
+        "yyyy-MMMM-dd HH:mm", "yyyy/MMMM/dd HH:mm", "yyyy MMMM dd HH:mm",
+        "dd-MMMM-yyyy HH:mm", "dd/MMMM/yyyy HH:mm", "dd MMMM yyyy HH:mm"
+    };
```
This is all very hard work! Maybe you can consider using [Regex](https://www.w3schools.com/java/java_regex.asp)?

deeyonn commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,114 @@
+     * @param fileDirectoryString Folder directory to put the storage in.
+     * @param fileName File name of the storage.
+     */
+    public Storage(String fileDirectoryString, String fileName) {
+        this.parser = new Parser();
+        File fileDirectory = new File(fileDirectoryString);
+        if (!fileDirectory.exists()) {
+            fileDirectory.mkdir();
+        }
+        this.filePath = fileDirectoryString + fileName;
+        File file = new File(this.filePath);
+        if (!file.exists()) {
+            try {
+                file.createNewFile();
+            } catch (IOException e) {
+                System.out.println(e.getMessage());
+            }
+        }
+    }
```
Maybe you can consider using a relative file path instead
```java
public Storage(String folderName) throws DukeException {
    String userHome = System.getProperty("user.home");
    this.storagePath = Paths.get(userHome, ".duke", "storage", folderName);
    try {
        Files.createDirectories(storagePath);
    } catch (UnsupportedOperationException | SecurityException | IOException e) {
        throw new DukeException (e.getMessage());
    }
}
```

deeyonn reviewed (commented):
Looks great to me! Keep up the good work!! 😄

nickeltea commented on src/main/java/bocil/Bocil.java:
```diff
@@ -0,0 +1,68 @@
+
+    private static final String NAME = "Bocil";
+    private static final String FILE_DIRECTORY_STRING = "./data/";
+    private static final String FILE_NAME = "duke.txt";
```
Good job following the coding standard for naming constants!

nickeltea commented on src/main/java/bocil/Bocil.java:
```diff
@@ -0,0 +1,68 @@
+    private final Parser parser;
+    private final Storage storage;
+    private final TaskList taskList;
+    private final Ui ui;
```
Are these also considered constants since they are final?

nickeltea commented on src/main/java/executor/Executor.java:
```diff
@@ -0,0 +1,189 @@
+ */
+public class Executor {
+
+    private final TaskList taskList;
```
Nice job following camelCase!

nickeltea commented on src/main/java/executor/Executor.java:
```diff
@@ -0,0 +1,189 @@
+     * @return Response of the program.
+     * @throws BocilException If the user input is not of the accepted format.
+     */
+    public String unmarkAsDone(String[] split) throws BocilException {
```
Well done following the coding standard by naming methods as verbs!

nickeltea commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,127 @@
+public class Parser {
+
+    /** Date patterns accepted by the parser. */
+    private static final String[] DATE_PATTERNS = {
```
Nice job using plural form!

JasonCP14 replied:
Thanks for the comments!

JasonCP14 committed:
Add basic GUI

JasonCP14 committed:
Add level 10 functionality

JasonCP14 committed:
Add JavaDocs for inherited method

JasonCP14 committed:
Personalize and optimize UI responses

JasonCP14 committed:
Add invalid data file formatting exception

JasonCP14 committed:
Add assertions

JasonCP14 committed:
Improve code quality

Code has lots of flaws in terms of code quality.

Modifying parts of codes will make it easier for other people to read
and understand it.

Let's,
* use guard clauses to make the happy path more prominent
* add comments to empty catch blocks
* add default cases for all switch-case statements

JasonCP14 committed:
Merge pull request #1 from JasonCP14/branch-A-Assertions

Add assertions

JasonCP14 committed:
Merge branch 'master' into branch-A-CodeQuality

JasonCP14 committed:
Merge pull request #2 from JasonCP14/branch-A-CodeQuality

Improve code quality

JasonCP14 committed:
Revert "Improve code quality"

JasonCP14 committed:
Merge pull request #3 from JasonCP14/revert-2-branch-A-CodeQuality

Revert "Improve code quality"

JasonCP14 committed:
Change code structure

JasonCP14 committed:
Add tags

JasonCP14 committed:
Refactor Executor methods

JasonCP14 committed:
Refactor redundant statements into Executor addNewTask()

JasonCP14 committed:
Add more personality

JasonCP14 committed:
Add background to application

JasonCP14 committed:
Add angry Bocil face to show invalid inputs

JasonCP14 committed:
Update User Guide
